pkg/controllers/corecluster/corecluster.go

```diff
@@ -0,0 +1,173 @@
+	Cluster  *clusterv1.Cluster
+	Infra    *configv1.Infrastructure
+	Platform configv1.PlatformType
+}
+
+// SetupWithManager sets the CoreClusterReconciler controller up with the given manager.
+func (r *CoreClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		For(r.Cluster).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the core cluster object for the openshift-cluster-api namespace.
+func (r *CoreClusterReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+	logger.V(1).Info("Reconciling core cluster")
+
+	// Still have to patch the condition onto a newly created cluster
+	cluster, err := r.ensureCoreCluster(ctx, r.Cluster.Name, r.Infra.Kind, logger)
+	if err != nil {
+		logger.Info("failed to create new core cluster.")
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.Create(ctx, cluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create Core Cluster: %w", err)
+	}
+
+	logger.Info(fmt.Sprintf("InfraCluster '%s/%s' successfully created", CAPINamespace, r.Infra.Status.InfrastructureName))
+
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil && !errors.IsNotFound(err) {
+		return ctrl.Result{}, fmt.Errorf("failed to get core cluster: %w", err)
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	clusterCopy := cluster.DeepCopy()
+
+	// Check if the Control Plane Initialized condition is set on an already existing cluster.
+	// If it's not, set it.
+	if conditions.Get(cluster, clusterv1.ControlPlaneInitializedCondition) == nil {
+		conditions.MarkTrue(cluster, clusterv1.ControlPlaneInitializedCondition)
+
+		patch := client.MergeFrom(clusterCopy)
+
+		isRequired, err := util.IsPatchRequired(cluster, patch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to check if patch required: %w", err)
+		}
+
+		if isRequired {
+			if err := r.Status().Patch(ctx, cluster, patch); err != nil {
+				return ctrl.Result{}, fmt.Errorf("unable to update core cluster status: %w", err)
+			}
+		}
+
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+	}
+
+	defer logger.V(1).Info("Finished reconciling core cluster")
+
+	return ctrl.Result{}, nil
+}
+
+// CreateCluster creates a cluster with the given name and returns the cluster object.
+func (r *CoreClusterReconciler) ensureCoreCluster(ctx context.Context, clusterName, infraClusterKind string, logger logr.Logger) (*clusterv1.Cluster, error) {
+	cluster := &clusterv1.Cluster{}
```

Comment:
You'll need to set the namespace and name for the cluster if you want to use it with `client.ObjectKeyFromObject`

Reply:
Otherwise you can use 
```
clusterKey := client.ObjectKey{
		Namespace: CAPINamespace,
		Name:      clusterName,
	}
```

Reply:
Missed that, thank you!